src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -65,9 +65,10 @@ protected function buildRoleMap()
                 }
 
                 $visited[] = $role;
-                $this->map[$main] = array_unique(array_merge($this->map[$main], $this->hierarchy[$role]));
+                $this->map[$main] = array_merge($this->map[$main], $this->hierarchy[$role]);
                 $additionalRoles = array_merge($additionalRoles, array_diff($this->hierarchy[$role], $visited));
             }
+            $this->map[$main] = array_unique($this->map[$main]);
         }
```

Comment:
>I assume without is the for-each you've added and not the array merges in the iteration?

Yes

Also we can replace `array_unique` with `array_keys` and write roles to the keys, it could be faster